### PR TITLE
Nb/feat/under annotated checks

### PR DESCRIPTION
**Pull request by Nadav-Barak**
Closes https://github.com/deepchecks/deepchecks/issues/2380

image

### CODE REVIEW

File: deepchecks/utils/abstracts/weak_segment_abstract.py [hunk 1]
```diff
@@ -291,7 +298,6 @@ def _format_partition_vec_for_display(self, partition_vec: np.array, feature_nam
 
     def _generate_check_result_value(self, weak_segments_df, cat_features: List[str], avg_score: float):
         """Generate a uniform format check result value for the different WeakSegmentsPerformance checks."""
-        pd.set_option('mode.chained_assignment', None)
```
Comment (noamzbr) [inline deepchecks/utils/abstracts/weak_segment_abstract.py:294]: This helps remove an annoying warning

File: deepchecks/utils/single_sample_metrics.py [hunk 2]
```diff
@@ -34,7 +34,7 @@ def calculate_neg_cross_entropy_per_sample(labels, probas: np.ndarray, model_cla
     if model_classes is not None:
         if any(x not in model_classes for x in labels):
             raise DeepchecksValueError(
-                f'Label observed values {sorted(labels.unique())} contain values '
+                f'Label observed values {sorted(np.unique(labels))} contain values '
```
Comment (noamzbr) [inline deepchecks/utils/single_sample_metrics.py:37]: make sure print is of a list and not a numpy array

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+# Copyright (C) 2021-2023 Deepchecks (https://www.deepchecks.com)
+#
+# This file is part of Deepchecks.
+# Deepchecks is distributed under the terms of the GNU Affero General
+# Public License (version 3 or later).
+# You should have received a copy of the GNU Affero General Public License
+# along with Deepchecks.  If not, see <http://www.gnu.org/licenses/>.
+# ----------------------------------------------------------------------------
+#
+"""Module of weak segments performance check."""
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:11]:
```suggestion
"""Module of the under annotated segment check."""
```

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        self.ignore_columns = ignore_columns
+        self.n_top_features = n_top_features
+        self.segment_minimum_size_ratio = segment_minimum_size_ratio
+        self.n_samples = n_samples
+        self.n_to_show = n_to_show
+        self.categorical_aggregation_threshold = categorical_aggregation_threshold
+
+    def run_logic(self, context: Context, dataset_kind) -> CheckResult:
+        """Run check."""
+        context.raise_if_token_classification_task(self)
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:52]: Please add a task to support both of these, defining a null label for them shouldn't be crazy hard.

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+        context.raise_if_multi_label_task(self)
+
+        text_data = context.get_data_by_kind(dataset_kind)
+        text_data = text_data.sample(self.n_samples, random_state=context.random_state)
+
+        features, cat_features = get_relevant_data_table(text_data, data_type=self.segment_by,
+                                                         columns=self.columns, ignore_columns=self.ignore_columns,
+                                                         n_top_features=self.n_top_features)
+
+        encoded_dataset = self._target_encode_categorical_features_fill_na(features, text_data.label,
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:62]: I would think you want to do categorical encoding to the label of isnull(text_data.label), not to the label itself. Would get better partition.

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+                                                   score_per_sample=score_per_sample,
+                                                   scorer_name='Annotation Ratio')
+
+        if len(weak_segments) == 0:
+            raise DeepchecksProcessError('Check was unable to find under annotated segments. This is expected if '
+                                         'your data is well annotated. If this is not the case, try increasing '
+                                         f'n_samples or supply more {self.segment_by}.')
+
+        check_result_value = self._generate_check_result_value(weak_segments, cat_features, avg_score)
+        display_msg = f'Showcasing intersections of {self.segment_by} with most under annotated segments.<br> The ' \
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:77]:
```suggestion
        display_msg = f'Showcasing intersections of {self.segment_by} that result in the most under annotated segments.<br> The ' \
```

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+                      'full list of under annotated segments can be observed in the check result value. '
+        display_fig = self._generate_scatter_plot_display(encoded_dataset.features_columns, score_per_sample,
+                                                          text_data.text, weak_segments, cat_features, avg_score)
+        return CheckResult(value=check_result_value, display=[display_msg, display_fig])
+
+    @staticmethod
+    def _get_box_boundaries(feature_data: pd.Series, segment: Tuple[float, float]) -> Tuple[float, float]:
+        lower = segment[0] if segment[0] != -np.inf else np.nanmin(feature_data)
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:85]: maybe use np.isfinite as it covers +- inf and also nans

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        annotated_data = sampled_data[is_annotated == 1]
+        not_annotated_data = sampled_data[is_annotated == 0]
+        for _, row in weak_segments.iterrows():
+            fig = go.Figure()
+            feature_1, feature_2 = row['Feature1'], row['Feature2']
+            feature_1_lower, feature_1_upper = self._get_box_boundaries(encoded_data[feature_1], row['Feature1 Range'])
+            if feature_2 != '':  # segment by two features
+                feature_2_lower, feature_2_upper = self._get_box_boundaries(encoded_data[feature_2],
+                                                                            row['Feature2 Range'])
+                hover_template = feature_1 + ': %{x}<br>' + feature_2 + ': %{y}<br>text: %{text}<br>Annotated: '
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:111]: Consider using bold `<b>` for the text before the ":"

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+                tab_title = f'{feature_1} vs {feature_2}'
+                msg = f'Under annotated segment contains samples with {feature_1} in range ' \
+                      f'{[format_number(feature_1_lower), format_number(feature_1_upper)]} and {feature_2} in range ' \
+                      f'{[format_number(feature_2_lower), format_number(feature_2_upper)]}.'
+            else:  # segment by one feature
+                feature_2 = 'virtual_col'
+                feature_2_lower = encoded_data['virtual_col'].min() * 1.3
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:118]: 0.25 * 1.3? Not better to have the final result in `jitter`?

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+                                     name='Not Annotated Samples'))
+
+            # Update figure layout
+            fig.update_layout(title=dict(
+                text=f'Under Annotated Segment ({row["% of Data"]}% of Data)<br><sub>Annotation ratio in segment ' \
+                     f'is {format_percent(row["Annotation Ratio"])} (vs. {format_percent(avg_score)}'
+                     f' in whole data)</sub>',
+                font=dict(size=24)),
+                xaxis_title=feature_1, yaxis_title=feature_2 if feature_2 != 'virtual_col' else '',
+                autosize=False, width=1000, height=600,
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:160]: If using constant size test on various displays (big screen, mac screen, 13'' screen (nir has one))

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+                font=dict(size=14),
+                plot_bgcolor='rgba(245, 245, 245, 1)',
+                xaxis=dict(gridcolor='rgba(200, 200, 200, 0.5)',
+                           zerolinecolor='rgba(200, 200, 200, 0.5)'),
+                yaxis=dict(gridcolor='rgba(200, 200, 200, 0.5)',
+                           zerolinecolor='rgba(200, 200, 200, 0.5)'))
+
+            display_tabs[tab_title] = [fig, f'Check ran on {encoded_data.shape[0]} data samples.<br>{msg}']
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:168]: maybe message before the number of samples?

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+
+        return DisplayMap(display_tabs)
+
+    def add_condition_segments_annotation_ratio_greater_than(self, threshold: float = 0.70):
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:172]: Add condition for "segment is less annotated than the avg annotation ratio by x"

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+        return self.add_condition(f'In all segments annotation ratio should be greater than '
+                                  f'{format_percent(threshold)}.', condition)
+
+
+class UnderAnnotatedPropertySegments(UnderAnnotatedSegments):
+    """Search for under annotated data segments.
+
+    The check is designed to help you easily identify under annotated segments of your data. The segments are
+    based on the text properties - which are features extracted from the text, such as "language" and
+    "number of words".
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:198]: link to property guide?

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+
+    In order to achieve this, the check trains several simple tree based models which try to predict the error of the
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:200]: update this text

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+                         categorical_aggregation_threshold=categorical_aggregation_threshold,
+                         **kwargs)
+
+
+class UnderAnnotatedMetaDataSegments(UnderAnnotatedSegments):
+    """Search for under annotated data segments.
+
+    The check is designed to help you easily identify under annotated segments of your data. The segments are
+    based on the metadata - which is data that is not part of the text, but is related to it,
+    such as "user_id" and "user_age".
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:248]: link to metadata guide?

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+
+    In order to achieve this, the check trains several simple tree based models which try to predict the error of the
```
Comment (noamzbr) [inline deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:250]: same comment

File: tests/nlp/checks/data_integrity/under_annotated_segments_test.py [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+def test_tweet_emotion_metadata_fully_annotated(tweet_emotion_train_test_textdata):
+    # Arrange
+    _, test = tweet_emotion_train_test_textdata
+    check = UnderAnnotatedMetaDataSegments().add_condition_segments_relative_performance_greater_than()
+
+    # Act & Assert
+    assert_that(
+        calling(check.run).with_args(test),
+        raises(DeepchecksProcessError, 'Check was unable to find under annotated segments. This is expected if '
+                                       'your data is well annotated. If this is not the case, try increasing '
```
Comment (noamzbr) [inline tests/nlp/checks/data_integrity/under_annotated_segments_test.py:76]: Can we test that given nulls in a certain metadata segment, the check finds a segment on the cross of the two correct metadata cols?

**Review by noamzbr — COMMENTED** [review]
Add to default suite.

File: deepchecks/utils/single_sample_metrics.py (hunk 2 above)
Comment (Nadav-Barak) [reply deepchecks/utils/single_sample_metrics.py:37]: its a list

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py (hunk 9 above)
Comment (Nadav-Barak) [reply deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:118]: no, it a display mechanism

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py (hunk 10 above)
Comment (Nadav-Barak) [reply deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:160]: That how it was defined in our plot files - see confusion matrix abstract, core/display.py
Anyhow could make it work with aspect ratio after spending 30-min and its not worth more

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py (hunk 11 above)
Comment (Nadav-Barak) [reply deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:168]: not sure what you meant
Anyhow it looks good under the tabs setting, see the display image

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py (hunk 12 above)
Comment (Nadav-Barak) [reply deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:172]: there is from abstract class

File: deepchecks/nlp/checks/data_integrity/under_annotated_segments.py (hunk 4 above)
Comment (Nadav-Barak) [reply deepchecks/nlp/checks/data_integrity/under_annotated_segments.py:52]: https://github.com/deepchecks/deepchecks/issues/2509

**Review by noamzbr — APPROVED** [review]

**Pull request merged**